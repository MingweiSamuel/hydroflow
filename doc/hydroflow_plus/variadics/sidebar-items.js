window.SIDEBAR_ITEMS = {"macro":["var_args","var_expr","var_type","variadic_trait"],"trait":["HomogenousVariadic","Split","UnrefVariadic","Variadic","VariadicExt"]};