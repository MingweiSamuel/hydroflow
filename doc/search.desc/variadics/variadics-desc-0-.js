searchState.loadedDescShard("variadics", 0, "Variadics\nThis as a variadic of exclusive (<code>mut</code>) references.\nThis as a variadic of references.\nCreates a new (longer) variadic type by appending <code>Suffix</code> …\nA variadic where all elements are the same type, <code>T</code>.\nIterator type returned by <code>into_iter</code>.\nIterator type returned by <code>Self::iter_any_mut</code>.\nIterator type returned by <code>Self::iter_any_ref</code>.\nThe number of items in this variadic (its length).\nThe reverse of this variadic type.\nHelper trait for splitting a variadic into two parts. …\nThe second part when splitting this variadic by <code>Prefix</code>.\nThe un-referenced variadic. Each item will have one layer …\nConvert from a variadic of references back into the …\nA variadic tuple list.\nExtension methods/types for <code>Variadic</code>s.\nConvert an exclusive (<code>mut</code>) reference to this variadic into …\nConvert a reference to this variadic into a variadic of …\nExtends this variadic value by appending <code>suffix</code> onto the …\nReturns a reference to an element.\nReturns an exclusive reference to an element.\nTurns this <code>HomogenousVariadic&lt;T&gt;</code> into an iterator of items …\nIterate this variadic as <code>&amp;mut dyn Any</code> exclusive references.\nIterate this variadic as <code>&amp;dyn Any</code> references.\nReverses this variadic value.\nSplits this variadic into two parts, first the <code>Prefix</code>, and …\nVariadic patterns macro.\nVariadic expressions (values) macro.\nVariadic types macro.\nThis macro generates a basic variadic trait where each …")