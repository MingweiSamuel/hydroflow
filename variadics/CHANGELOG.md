# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.0.5 (2024-07-23)

### New Features

 - <csr-id-20080cb7ceb5b5d3ba349dfd822a37288e40add6/> add traits for dealing with variadics of references
   Renames some traits, but not a breaking change since there hasn't been a
   release that includes those traits.
 - <csr-id-b92dfc7460c985db6935e79d612f42b9b87e746f/> add `iter_any_ref` and `iter_any_mut` to `VariadicsExt`
   Depends on #1241
   
   This isn't needed for the current GHT implementation, but is useful in
   general
 - <csr-id-1a6228f2db081af68890e2e64b3a91f15dd9214f/> add traits for referencing variadics
   This adds a way to convert a reference to a variadic into a variadic of
   references. I.e. `&var_expr!(a, b, c) -> var_expr!(&a, &b, &c)`

### Bug Fixes

 - <csr-id-bbef0705d509831415d3bb5ce003116af06b6ffb/> `EitherRefVariadic` is `Variadic`
 - <csr-id-c70114d836e5bc36e2104188867e548e90ab38f4/> fix `HomogenousVariadic` `get` and `get_mut` only returning `None`

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 5 commits contributed to the release over the course of 53 calendar days.
 - 143 days passed between releases.
 - 5 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 5 unique issues were worked on: [#1241](https://github.com/hydro-project/hydroflow/issues/1241), [#1245](https://github.com/hydro-project/hydroflow/issues/1245), [#1324](https://github.com/hydro-project/hydroflow/issues/1324), [#1325](https://github.com/hydro-project/hydroflow/issues/1325), [#1352](https://github.com/hydro-project/hydroflow/issues/1352)

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **[#1241](https://github.com/hydro-project/hydroflow/issues/1241)**
    - Add traits for referencing variadics ([`1a6228f`](https://github.com/hydro-project/hydroflow/commit/1a6228f2db081af68890e2e64b3a91f15dd9214f))
 * **[#1245](https://github.com/hydro-project/hydroflow/issues/1245)**
    - Add `iter_any_ref` and `iter_any_mut` to `VariadicsExt` ([`b92dfc7`](https://github.com/hydro-project/hydroflow/commit/b92dfc7460c985db6935e79d612f42b9b87e746f))
 * **[#1324](https://github.com/hydro-project/hydroflow/issues/1324)**
    - Add traits for dealing with variadics of references ([`20080cb`](https://github.com/hydro-project/hydroflow/commit/20080cb7ceb5b5d3ba349dfd822a37288e40add6))
 * **[#1325](https://github.com/hydro-project/hydroflow/issues/1325)**
    - Fix `HomogenousVariadic` `get` and `get_mut` only returning `None` ([`c70114d`](https://github.com/hydro-project/hydroflow/commit/c70114d836e5bc36e2104188867e548e90ab38f4))
 * **[#1352](https://github.com/hydro-project/hydroflow/issues/1352)**
    - `EitherRefVariadic` is `Variadic` ([`bbef070`](https://github.com/hydro-project/hydroflow/commit/bbef0705d509831415d3bb5ce003116af06b6ffb))
</details>

## 0.0.4 (2024-03-02)

<csr-id-5a451ac4ae75024153a06416fc81d834d1fdae6f/>
<csr-id-7103e77d0da1d73f1c93fcdb260b6a4c9a18ff66/>
<csr-id-b4683450a273d510a11338f07920a5558033b31f/>

### Chore

 - <csr-id-5a451ac4ae75024153a06416fc81d834d1fdae6f/> prep for 0.0.4 release
 - <csr-id-7103e77d0da1d73f1c93fcdb260b6a4c9a18ff66/> update pinned rust to 2024-04-24

### Style

 - <csr-id-b4683450a273d510a11338f07920a5558033b31f/> fix dead code lint

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 4 commits contributed to the release over the course of 3 calendar days.
 - 31 days passed between releases.
 - 3 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 0 issues like '(#ID)' were seen in commit messages

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **Uncategorized**
    - Release hydroflow_lang v0.6.0, hydroflow_datalog_core v0.6.0, hydroflow_datalog v0.6.0, hydroflow_macro v0.6.0, lattices v0.5.3, variadics v0.0.4, pusherator v0.0.5, hydroflow v0.6.0, stageleft v0.2.0, hydroflow_plus v0.6.0, hydro_deploy v0.6.0, hydro_cli v0.6.0, hydroflow_plus_cli_integration v0.6.0, safety bump 7 crates ([`09ea65f`](https://github.com/hydro-project/hydroflow/commit/09ea65fe9cd45c357c43bffca30e60243fa45cc8))
    - Prep for 0.0.4 release ([`5a451ac`](https://github.com/hydro-project/hydroflow/commit/5a451ac4ae75024153a06416fc81d834d1fdae6f))
    - Fix dead code lint ([`b468345`](https://github.com/hydro-project/hydroflow/commit/b4683450a273d510a11338f07920a5558033b31f))
    - Update pinned rust to 2024-04-24 ([`7103e77`](https://github.com/hydro-project/hydroflow/commit/7103e77d0da1d73f1c93fcdb260b6a4c9a18ff66))
</details>

## 0.0.3 (2024-01-29)

<csr-id-1b555e57c8c812bed4d6495d2960cbf77fb0b3ef/>
<csr-id-7e65a08711775656e435e854777c5f089dd31a05/>

### Chore

 - <csr-id-1b555e57c8c812bed4d6495d2960cbf77fb0b3ef/> manually set lockstep-versioned crates (and `lattices`) to version `0.5.1`
   Setting manually since
   https://github.com/frewsxcv/rust-crates-index/issues/159 is messing with
   smart-release

### Refactor

 - <csr-id-7e65a08711775656e435e854777c5f089dd31a05/> Improvements prepping for release
   - Adds the "spread"/"splat" `...` syntax to the three variadics macros.
   - Adds `#[sealed]` traits.
   - Adds testing of error messages.
   - Improves docs: `README.md` and Rust docs.

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 3 commits contributed to the release over the course of 37 calendar days.
 - 253 days passed between releases.
 - 2 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 1 unique issue was worked on: [#974](https://github.com/hydro-project/hydroflow/issues/974)

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **[#974](https://github.com/hydro-project/hydroflow/issues/974)**
    - Improvements prepping for release ([`7e65a08`](https://github.com/hydro-project/hydroflow/commit/7e65a08711775656e435e854777c5f089dd31a05))
 * **Uncategorized**
    - Release hydroflow_cli_integration v0.5.1, hydroflow_lang v0.5.1, hydroflow_datalog_core v0.5.1, hydroflow_datalog v0.5.1, hydroflow_macro v0.5.1, lattices v0.5.1, variadics v0.0.3, pusherator v0.0.4, hydroflow v0.5.1, stageleft_macro v0.1.0, stageleft v0.1.0, hydroflow_plus v0.5.1, hydro_deploy v0.5.1, hydro_cli v0.5.1 ([`478aebc`](https://github.com/hydro-project/hydroflow/commit/478aebc8fee2aa78eab86bd386322db1c70bde6a))
    - Manually set lockstep-versioned crates (and `lattices`) to version `0.5.1` ([`1b555e5`](https://github.com/hydro-project/hydroflow/commit/1b555e57c8c812bed4d6495d2960cbf77fb0b3ef))
</details>

## 0.0.2 (2023-05-21)

<csr-id-5a3c2949653685de1e33cf7412057a70880283df/>

### Style

 - <csr-id-5a3c2949653685de1e33cf7412057a70880283df/> rustfmt format code comments

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 2 commits contributed to the release.
 - 24 days passed between releases.
 - 1 commit was understood as [conventional](https://www.conventionalcommits.org).
 - 1 unique issue was worked on: [#660](https://github.com/hydro-project/hydroflow/issues/660)

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **[#660](https://github.com/hydro-project/hydroflow/issues/660)**
    - Rustfmt format code comments ([`5a3c294`](https://github.com/hydro-project/hydroflow/commit/5a3c2949653685de1e33cf7412057a70880283df))
 * **Uncategorized**
    - Release hydroflow_cli_integration v0.0.1, hydroflow_lang v0.0.1, hydroflow_datalog_core v0.0.1, hydroflow_datalog v0.0.1, hydroflow_macro v0.0.1, lattices v0.1.0, variadics v0.0.2, pusherator v0.0.1, hydroflow v0.0.2 ([`809395a`](https://github.com/hydro-project/hydroflow/commit/809395acddb78949d7a2bf036e1a94972f23b1ad))
</details>

## 0.0.1 (2023-04-26)

### Commit Statistics

<csr-read-only-do-not-edit/>

 - 4 commits contributed to the release over the course of 133 calendar days.
 - 0 commits were understood as [conventional](https://www.conventionalcommits.org).
 - 1 unique issue was worked on: [#617](https://github.com/hydro-project/hydroflow/issues/617)

### Commit Details

<csr-read-only-do-not-edit/>

<details><summary>view details</summary>

 * **[#617](https://github.com/hydro-project/hydroflow/issues/617)**
    - Update `Cargo.toml`s for publishing ([`a78ff9a`](https://github.com/hydro-project/hydroflow/commit/a78ff9aace6771787c2b72aad83be6ad8d49a828))
 * **Uncategorized**
    - Setup release workflow ([`108d0e9`](https://github.com/hydro-project/hydroflow/commit/108d0e933a08b183c4dadf8c3499e4946696e263))
    - Rename variadics/tuple_list macros ([`91d37b0`](https://github.com/hydro-project/hydroflow/commit/91d37b022b1cd0ed590765c40ef43244027c8035))
    - Rename pkg `type_list` -> `variadics` ([`50e7361`](https://github.com/hydro-project/hydroflow/commit/50e7361709cd34fd0e1cbf0c9a9f79343ee9c2e2))
</details>

