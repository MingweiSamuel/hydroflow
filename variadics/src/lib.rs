#![doc = include_str!("../README.md")]
#![warn(missing_docs)]

use sealed::sealed;

/// Macro creating a variadic tuple value from a list of expressions.
///
/// ### Examples
///
/// Create a variadic tuple value:
/// ```rust
/// use variadics::var;
///
/// let list = var!(10, false, "foo");
///
/// assert_eq!(list, (10, (false, ("foo", ()))),)
/// ```
///
/// Although this can be used as a pattern unpack tuples, [`varg!`] should be used instead:
/// ```
/// # use variadics::*;
/// // Ok...
/// let var!(a, b, c) = var!(10, false, "foo");
/// // Better:
/// let varg!(a, b, c) = var!(10, false, "foo");
///
/// assert_eq!(a, 10);
/// assert_eq!(b, false);
/// assert_eq!(c, "foo");
/// ```
///
/// The "spread" (or "splat") syntax `...` can be used to concatenate variadics together:
/// ```rust
/// # use variadics::var;
/// let list_a = var!(0.5, "foo");
/// let list_b = var!(-5, false);
/// // Spread syntax:
/// let list_c = var!(...list_a, ...list_b, "bar");
/// // Equals `var!(0.5, "foo", -5, false, "bar)`.
/// ```
#[macro_export]
macro_rules! var {
    () => ( () );

    (...$a:ident $(,)? ) => ( $a );
    (...$a:expr  $(,)? ) => ( $a );
    (...$a:ident, $( $b:tt )+) => ( $crate::VariadicExt::extend($a, $crate::var!( $( $b )* )) );
    (...$a:expr,  $( $b:tt )+) => ( $crate::VariadicExt::extend($a, $crate::var!( $( $b )* )) );

    ($a:ident $(,)? ) => ( ($a, ()) );
    ($a:expr  $(,)? ) => ( ($a, ()) );
    ($a:ident, $( $b:tt )+) => ( ($a, $crate::var!( $( $b )* )) );
    ($a:expr,  $( $b:tt )+) => ( ($a, $crate::var!( $( $b )* )) );
}

/// Macro creating a variadic tuple type from a list of types.
///
/// `var!` can be used to define simple types but will result in confusing errors for more
/// complex types. Use this macro, `Var!` instead.
///
/// # Examples
///
/// ```rust
/// # use std::collections::HashMap;
/// use variadics::{var, Var};
///
/// // A simple variadic type. Although `var!` would work in this case, it cannot handle
/// // more complex types i.e. ones with generics.
/// let list: Var!(i32, bool, String) = Default::default();
///
/// // A more complex type:
/// let list: Var!(
///     &'static str,
///     HashMap<i32, i32>,
///     <std::vec::Vec<bool> as IntoIterator>::Item,
/// ) = var!("foo", HashMap::new(), false);
/// ```
///
/// Unfortunately, expressions and types cannot be handled using the same macro due to the
/// undefeated [bastion of the turbofish](https://github.com/rust-lang/rust/blob/7fd15f09008dd72f40d76a5bebb60e3991095a5f/src/test/ui/parser/bastion-of-the-turbofish.rs).
#[macro_export]
macro_rules! Var {
    () => ( () );

    (...$a:ty $(,)? ) => ( $a );
    (...$a:ty, $( $b:tt )+) => ( <$a as $crate::VariadicExt::Extend<$crate::Var!( $( $b )* )>>);

    ($a:ty $(,)? ) => ( ($a, ()) );
    ($a:ty, $( $b:tt )+) => ( ($a, $crate::Var!( $( $b )* )) );
}

/// Macro for [pattern-matching](https://doc.rust-lang.org/reference/patterns.html) with variadic
/// tuples. This is used for function arguments, as well as in `match`, `if/while let ...`,
/// `let ... else`, and `for` expressions.
///
/// Although it may somtimes be possible to use `var!` in place of this macro, doing so may
/// cause confusing errors.
///
/// # Examples
///
/// ```rust
/// use variadics::{var, varg, Var};
///
/// fn my_fn(varg!(a, b, c): Var!(usize, &str, bool)) {
///     println!("{} {} {}", a, b, c);
/// }
/// my_fn(var!(12, "hello", false));
/// ```
///
/// ```rust
/// use variadics::{var, varg};
///
/// let val = var!(true, Some("foo"), 2);
/// if let varg!(true, Some(item), 0..=3) = val {
///     println!("{}", item);
/// } else {
///     unreachable!();
/// }
/// ```
///
/// ```rust
/// # use variadics::{var, varg};
/// match var!(true, Some(100), 5) {
///     varg!(false, _, _) => unreachable!(),
///     varg!(true, None, _) => unreachable!(),
///     varg!(true, Some(0..=10), _) => unreachable!(),
///     varg!(true, Some(a), b) => println!("{} {}", a, b),
/// }
#[macro_export]
macro_rules! varg {
    () => ( () );

    (...$a:pat $(,)? ) => ( $a );
    (...$a:ty, $( $b:tt )+) => ( ::core::compile_error!("`varg!` can only have the `...` spread syntax on the last field.") );

    ($a:pat $(,)? ) => ( ($a, ()) );
    ($a:pat, $( $b:tt )+) => ( ($a, $crate::varg!( $( $b )* )) );
}

/// This macro generates a basic variadic trait where each element must fulfill the `where` clause.
///
/// ```rust
/// use variadics::{var, variadic_trait};
///
/// variadic_trait! {
///     /// A variadic list of `Debug` items.
///     pub variadic<Item> DebugList where Item: std::fmt::Debug {}
/// }
///
/// let x = &var!(1, "hello", 5.6);
/// let _: &dyn DebugList = x;
/// println!("{:?}", x);
/// ```
///
/// This uses a special syntax similar to traits, but with the `trait` keyword replaced with
/// `variadic<T>` where `T` is the generic parameter name for each item in the variadic list. `T`
/// can be changed to any valid generic identifier. The bounds on `T` must be put in the where
/// clause; they cannot be expressed directly-- `variadic<T: Clone>` is invalid.
///
/// For now this can only create traits which bounds the `Item`s and cannot have associated
/// methods. This means the body of the variadic trait must be empty. But in the future this
/// declarative macro may be converted into a more powerful procedural macro with associated
/// method support.
#[macro_export]
macro_rules! variadic_trait {
    (
        $( #[$( $attrs:tt )*] )*
        $vis:vis variadic<$item:ident> $name:ident $( $clause:tt )*
    ) => {
        $( #[$( $attrs )*] )*
        $vis trait $name: $crate::Variadic {}
        $( #[$( $attrs )*] )*
        impl $name for $crate::Var!() {}
        $( #[$( $attrs )*] )*
        impl<$item, __Rest: $name> $name for $crate::Var!($item, ...__Rest) $( $clause )*
    };
}

/// A variadic tuple list.
///
/// This is a sealed trait, implemented only for `(Item, Rest) where Rest: Variadic` and `()`.
#[sealed]
pub trait Variadic {}
#[sealed]
impl<Item, Rest> Variadic for (Item, Rest) where Rest: Variadic {}
#[sealed]
impl Variadic for () {}

/// Extension methods/types for [`Variadic`]s.
///
/// This is a sealed trait.
#[sealed]
pub trait VariadicExt: Variadic {
    /// The number of items in this variadic (its length).
    const LEN: usize;

    /// Creates a new (longer) variadic type by appending `Suffix` onto the end of this variadc.
    type Extend<Suffix>: VariadicExt
    where
        Suffix: VariadicExt;
    /// Extends this variadic value by appending `suffix` onto the end.
    fn extend<Suffix>(self, suffix: Suffix) -> Self::Extend<Suffix>
    where
        Suffix: VariadicExt;

    /// The reverse of this variadic type.
    type Reverse: VariadicExt;
    /// Reverses this variadic value.
    fn reverse(self) -> Self::Reverse;
}
#[sealed]
impl<Item, Rest> VariadicExt for (Item, Rest)
where
    Rest: VariadicExt,
{
    const LEN: usize = 1 + Rest::LEN;

    type Extend<Suffix> = (Item, Rest::Extend<Suffix>) where Suffix: VariadicExt;
    fn extend<Suffix>(self, suffix: Suffix) -> Self::Extend<Suffix>
    where
        Suffix: VariadicExt,
    {
        let (item, rest) = self;
        (item, rest.extend(suffix))
    }

    type Reverse = <Rest::Reverse as VariadicExt>::Extend<(Item, ())>;
    fn reverse(self) -> Self::Reverse {
        let (item, rest) = self;
        rest.reverse().extend((item, ()))
    }
}
#[sealed]
impl VariadicExt for () {
    const LEN: usize = 0;

    type Extend<Suffix> = Suffix where Suffix: VariadicExt;
    fn extend<Suffix>(self, suffix: Suffix) -> Self::Extend<Suffix>
    where
        Suffix: VariadicExt,
    {
        suffix
    }

    type Reverse = ();
    fn reverse(self) -> Self::Reverse {}
}

/// A variadic where all elements are the same type, `T`.
///
/// This is a sealed trait.
#[sealed]
pub trait HomogenousVariadic<T>: Variadic {
    /// Returns a reference to an element.
    fn get(&self, i: usize) -> Option<&T>;
    /// Returns an exclusive reference to an element.
    fn get_mut(&mut self, i: usize) -> Option<&mut T>;

    /// Iterator type returned by `into_iter`.
    type IntoIter: Iterator<Item = T>;
    /// Turns this `HomogenousVariadic<T>` into an iterator of items `T`.
    fn into_iter(self) -> Self::IntoIter;
}
#[sealed]
impl<T> HomogenousVariadic<T> for () {
    fn get(&self, _i: usize) -> Option<&T> {
        None
    }
    fn get_mut(&mut self, _i: usize) -> Option<&mut T> {
        None
    }

    type IntoIter = std::iter::Empty<T>;
    fn into_iter(self) -> Self::IntoIter {
        std::iter::empty()
    }
}
#[sealed]
impl<T, Rest> HomogenousVariadic<T> for (T, Rest)
where
    Rest: HomogenousVariadic<T>,
{
    fn get(&self, i: usize) -> Option<&T> {
        let (item, rest) = self;
        if i == 0 {
            Some(item)
        } else {
            rest.get(i)
        }
    }
    fn get_mut(&mut self, i: usize) -> Option<&mut T> {
        let (item, rest) = self;
        if i == 0 {
            Some(item)
        } else {
            rest.get_mut(i)
        }
    }

    type IntoIter = std::iter::Chain<std::iter::Once<T>, Rest::IntoIter>;
    fn into_iter(self) -> Self::IntoIter {
        let (item, rest) = self;
        std::iter::once(item).chain(rest.into_iter())
    }
}

/// Helper trait for splitting a variadic into two parts. `Prefix` is the first part, everything
/// after is the `Suffix` or second part.
///
/// This is a sealed trait.
#[sealed]
pub trait Split<Prefix>: Variadic
where
    Prefix: Variadic,
{
    /// The second part when splitting this variadic by `Prefix`.
    type Suffix: Variadic;
    /// Splits this variadic into two parts, first the `Prefix`, and second the `Suffix`.
    fn split(self) -> (Prefix, Self::Suffix);
}
#[sealed]
impl<Item, Rest, PrefixRest> Split<(Item, PrefixRest)> for (Item, Rest)
where
    PrefixRest: Variadic,
    Rest: Split<PrefixRest>,
{
    type Suffix = <Rest as Split<PrefixRest>>::Suffix;
    fn split(self) -> ((Item, PrefixRest), Self::Suffix) {
        let (item, rest) = self;
        let (prefix_rest, suffix) = rest.split();
        ((item, prefix_rest), suffix)
    }
}
#[sealed]
impl<Rest> Split<()> for Rest
where
    Rest: Variadic,
{
    type Suffix = Rest;
    fn split(self) -> ((), Self::Suffix) {
        ((), self)
    }
}

#[cfg(test)]
mod test {
    use super::*;

    type MyList = Var!(u8, u16, u32, u64);
    type MyPrefix = Var!(u8, u16);

    type MySuffix = <MyList as Split<MyPrefix>>::Suffix;

    #[allow(dead_code)]
    const _: MySuffix = var!(0_u32, 0_u64);

    #[test]
    #[allow(clippy::let_unit_value)]
    fn test_basic_expr() {
        let _ = var!();
        let _ = var!(1);
        let _ = var!(1, "b",);
        let _ = var!("a",);
        let _ = var!(false, true, 1 + 2);
    }

    variadic_trait! {
        /// Variaidic list of futures.
        pub variadic<F> FuturesList where F: std::future::Future {}
    }
}
