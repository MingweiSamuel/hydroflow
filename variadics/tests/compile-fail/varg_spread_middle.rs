use variadics::*;

fn main() {
    let varg!(a, ...b, c) = var!(1, 2.0, "three", false);
}