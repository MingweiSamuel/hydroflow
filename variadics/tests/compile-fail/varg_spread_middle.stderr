error: `varg!` can only have the `...` spread syntax on the last field.
 --> tests/compile-fail/varg_spread_middle.rs:4:9
  |
4 |     let varg!(a, ...b, c) = var!(1, 2.0, "three", false);
  |         ^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `$crate::varg` which comes from the expansion of the macro `varg` (in Nightly builds, run with -Z macro-backtrace for more info)
